trie-db/src/iterator.rs

```diff
@@ -77,30 +76,31 @@ impl<'a, L: TrieLayout> TrieDBNodeIterator<'a, L> {
 			trail: Vec::with_capacity(8),
 			key_nibbles: NibbleVec::new(),
 		};
-		db.root_data().and_then(|root_data| {
-			r.descend(&root_data, Some(db.root().clone())).map(|_| ())
-		})?;
+		let (root_node, root_hash) = db.get_raw_or_lookup(db.root().as_ref(), EMPTY_PREFIX)?;
+		r.descend(root_node, root_hash);
 		Ok(r)
 	}
 
 	fn seek(
 		&mut self,
-		mut node_data: DBValue,
+		node_data: DBValue,
 		mut node_hash: Option<TrieHash<L>>,
 		key: NibbleSlice,
 	) -> Result<(), TrieHash<L>, CError<L>> {
+		let mut node = OwnedNode::new::<L::Hash, L::Codec>(node_data)
+			.map_err(|e| Box::new(TrieError::DecoderError(node_hash.unwrap_or_default(), e)))?;
 		let mut partial = key;
 		let mut full_key_nibbles = 0;
 		loop {
-			let (next_node_data, next_node_hash) = {
-				let node = self.descend(&node_data, node_hash)?;
+			let (next_node, next_node_hash) = {
+				self.descend(node, node_hash);
 				let crumb = self.trail.last_mut()
 					.expect(
 						"descend_into_node pushes a crumb onto the trial; \
 						thus the trail is non-empty; qed"
 					);
 
-				match node {
+				match crumb.node.node() {
```

Comment:
To get performance improvement, the number of 'node' call would probably need to be reduce, or maybe use OwnedNode<&'a[u8]> in crumb (not really sure).
Owned<&'a[u8]> for zero copy proof check, is also a bit of a challenge regarding lifetime management.

Reply:
Refactored this method to significantly reduce the number of calls to `node()`.